Mind Machine

GreatBigFailure opened the pull request:
**Updated sprites thanks to Megautham who told me not to compliment him too much**
They look really fantastic and I'm very grateful. 

The Mind Machine is Nanotrasen's latest experiment into bluespace technology. After one too many attacks by mind swapping Wizard's and their apprentices, not to mention compensating the families of employees catbeasted by a staff of change, they've developed this """"safe"""" and easy way to fix any mind related problem possible.

![bettermindmachien](https://user-images.githubusercontent.com/64218965/84706617-4851db00-af2b-11ea-966e-dd67d989dc98.png)


**What it can do**
The mind machine can swap the minds of any creature placed inside of it. This includes higher minded beings like employees, as well as lower minded like carp. It can even put higher minds into the carp. 
Mind swaps come with two modes: The normal recommended mode in which a supervisor scans and watches over the swap process, and the non-liable delayed scan mode in which an employee may hit the button then quickly jump into a pod himself.

**WARNING:** Supervision of swaps is highly recommended. Nanotrasen takes no responsibility for accidents occurring from the "delayed swap" mode

**How it works**
After charging the bluespace conduit by allowing it to assimilate a bluespace crystal just place both targets into a pod each, hit scan, then hit swap. The two minds will safely and comfortably travel between the pods and into the opposite body, wow!

![mmUI](https://user-images.githubusercontent.com/64218965/83907382-50dc3180-a733-11ea-88c8-4b618a345956.png)

**WARNING:** Please ensure pod is supplied with ample power during transfer process. Loss of power may result in altered mental transit location or non-liable brain damage.

**WARNING** Ensure only those wishing to participate in the transfer enter the pods. One occupant per pod. Any mice, diona, carp, chickens, cats, snakes, or other pets should wait for their owners outside. Failure to do so may result in miscalculation in transfer.

![mindtheflybetter](https://user-images.githubusercontent.com/64218965/84707037-feb5c000-af2b-11ea-9866-35a63b53d646.png)

**This PR: What's done**

The machine and all its parts are fully functional. It's obtainable via research using the nantrasen cargo disk. The circuits are under medical machines as that is its intended, legal, and exclusive use. 
This let's you:
- Play as a carp, Ian, spider queen, etc
- Make Ian think he's a wizard, carp, or zombie while making the carp you swapped him with your loyal, glass destroying pet
- Put people dying of the vampire spicy touch into a monkey man instead of cloning them, assuming they're still alive when you do it. 
- With enough upgrades, put a cyborg or the AI into Runtime for advanced silicon bully

It doesn't let you:
- Control the roboduck

Edit: Added the ability to struggle out of the machine if you're trapped. Either intentionally or due to bugs. Forces the machine to reset itself. Should now be perfectly safe.

:cl:
 * rscadd: Nanotrasen has created a prototype Mind Transfer Machine available for brave employees and their pets to test.

Irradiation commented:
This is a horrible fucking idea that treads upon the Admins' own territory and may or may not end in silly amounts of griff.

MERGE THIS C*LLABS.

BomberBro commented:
Very nice, only criticism I have is that the sprites look out of place compared to the other medbay/science machines, maybe needs a darker outline on some parts.

GreatBigFailure commented:
To save the collabs time, reducing the review/comment process from 9 hours to 8 hours 59 minutes:
You can't trap players in laz balls
Laz injectors cannot convert mob'd players into traitors when used by a syndy
Converting two players into carp of opposite genders doesn't result in one exploding into baby carp after a few minutes, unfortunately
Swapping into a syndicate mob doesn't make you a traitor, same with wizard and the like

GreatBigFailure commented:
> Very nice, only criticism I have is that the sprites look out of place compared to the other medbay/science machines, maybe needs a darker outline on some parts.

I'm not in love with the sprites. On their own I really like the pods and their animations but you're definitely right about them not fitting. I tried a darker version but it looked pretty bad. I want them to stay a The Fly reference since that was the whole idea. 
![theflyposter](https://user-images.githubusercontent.com/64218965/83910469-9a7b4b00-a738-11ea-8f36-7e19e500925a.jpg)

If a better spriter than me wants to take a crack at it I'd be glad, but I think this is the best I can do for now.

gurfan commented:
if you throw a simple creature in both pods, will the behavior of both creatures be swapped? 

ex. swap a spider with a corgi, the corgi will now attack people

GreatBigFailure commented:
> if you throw a simple creature in both pods, will the behavior of both creatures be swapped?
> 
> ex. swap a spider with a corgi, the corgi will now attack people

What it does is swap their faction, emotes, speak chance, and name. I tried to expand past that but for now that's all I can do. For some creatures it does make them hostile, for some it doesn't. Ian won't attack you for example. It will make hostile creatures friendly similar to laz/gold slime extracts.
Once I'm more knowledgeable I'll go back and see if I can do things like add attack procs to creatures, like Ian, without them. For now I'm torturing the collab's enough though.

DrSnips commented:
>Anom. AND NTE research

I'm not saying it's useless however you might as well not even added to the game at this point because that's not one but two points of NEVER EVER research.

GreatBigFailure commented:
> > Anom. AND NTE research
> 
> I'm not saying it's useless however you might as well not even added to the game at this point because that's not one but two points of NEVER EVER research.

It only requires like 2 spoons worth of anomaly and NT just needs a quick trip to cargo. There's been a lot of xenoarchs lately and if someone wants the machine they can just get the disk. I want it to be something uncommon, not super rare, just not an every shift necessity.
If it's never touched, or people really want me to, I'll lower the requirements though.

small-boss2 commented:
> It only requires like 2 spoons worth of anomaly and NT just needs a quick trip to cargo. There's been a lot of xenoarchs lately and if someone wants the machine they can just get the disk.

You underestimate scientists. They'll never do special research levels for things that won't help them powergame. Keeping it locked behind NTE research AND anomaly will make people forget about this machine in like a week due to nobody going through the effort to print it. Choose one, or go with something like requiring bluespace and/or medical research.

GreatBigFailure commented:
> You underestimate scientists. They'll never do special research levels for things that won't help them powergame. Keeping it locked behind NTE research AND anomaly will make people forget about this machine in like a week due to nobody going through the effort to print it. Choose one, or go with something like requiring bluespace and/or medical research.

I'll do some testing on making the bluespace/medical higher. I still think the requirements are actually pretty low. I'd keep it on the nano disk and remove anomaly then I guess, I want to buff that disk. Still not sure about lowering it but I'll keep getting more opinions and see how I can shift it around. Also considering adding another part to the construction requirements. Add that to the nano-disk but remove the circuits from it. I kind of want a fancy sounding item on the protolathe nano menu so that people see it and want to make it. No one can stomach the clusterfuck circuitboard menu.

GreatBigFailure commented:
Made vampires, cultists, changelings, and malf AIs unswappable. I didn't test the last one but I just sort of assumed it was a really bad idea. 
Added a tin foil hat interaction 
Added another thing
Menagerie doesn't seem to break everything but I only tested with 2 minds so collabs please consider it.

Caboozles commented:
Wouldn’t this act as a metacheck?

Unrelatedly, how does the game handle sacrifice targets, or the such? Is it the mind or the body that they want?

DrSnips commented:
I feel like for the time to get resources and research for the machine and if you haven't sufficiently figured out who is or isn't an antag you earned it. Also I have high doubt that security is going to set up one of these just the check every single person on the station.

GreatBigFailure commented:
>Wouldn’t this act as a metacheck?
That's originally one reason I was trying so hard to fix the spell issue but I just don't think I'll be able to. If anyone can then I'll remove the vampire/cult block. Keeping the malf though. I didn't even test that but that's for sure a bad idea. 

>I feel like for the time to get resources and research for the machine and if you haven't sufficiently figured out who is or isn't an antag you earned it
Also this though. Luckily both vamp and cult already have way, way easier detection methods so this isn't too crippling.

GreatBigFailure commented:
@Zth-- I'll update the wiki as soon as/if it's merged. Wouldn't be a long article but enough of a time investment I'd rather wait for confirmation.

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	name = "Mind Machine"
+	icon = 'icons/obj/mind_machine.dmi'
+	density = 1
+	anchored = 1.0
```
anchored is a binary value, drop that .0

PJB3005 replied:
More like use real boolean constants:

```suggestion
	anchored = TRUE
```

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	idle_power_usage = 50
+	active_power_usage = 2000
+	light_range_on = 1
+	light_power_on = 1
```
duplicated var

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	var/obj/machinery/mind_machine/mind_machine_pod/connectOne //All "variableNameOne" refer to information about the connectOne pod
+	var/obj/machinery/mind_machine/mind_machine_pod/connectTwo //Likewise for variableNameTwo for connectTwo
+	var/podsConnected = FALSE
+	var/occupantOne = null
```
typecasting the occupant to at least /mob/living would be nice

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	var/obj/machinery/mind_machine/mind_machine_pod/connectTwo //Likewise for variableNameTwo for connectTwo
+	var/podsConnected = FALSE
+	var/occupantOne = null
+	var/occupantTwo = null
```
this one too

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	var/podsConnected = FALSE
+	var/occupantOne = null
+	var/occupantTwo = null
+	var/occupantNameOne = "None"
```
you wouldn't need this after the occupant is typecast

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	var/occupantOne = null
+	var/occupantTwo = null
+	var/occupantNameOne = "None"
+	var/occupantNameTwo = "None"
```
ditto

PJB3005 replied:
See the problem with all these vars down here is that while they *are* completely inferrable from the contained mobs, it's kind of non-trivial I guess?

I would prefer if this data wasn't duplicated and the data was pulled from the mobs in `ui_interact()` instead to reduce duplication though, yeah.

Also makes it more clear what this is used for and the user-intended strings are then located in only `ui_interact()`

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+/obj/machinery/mind_machine/mind_machine_hub/Destroy()
+	podsConnected = null
+	currentlySwapping = null
+	lockedPods = FALSE
```
is this really necessary

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	podsConnected = null
+	currentlySwapping = null
+	lockedPods = FALSE
+	occupantScan = FALSE
```
same

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	occupantScan = FALSE
+	occupantOne = null
+	occupantTwo = null
+	swapProgress = 0
```
i mean you're destroying the thing...

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	occupantOne = null
+	occupantTwo = null
+	swapProgress = 0
+	if(connectOne || connectTwo || podsConnected)
```
making this a mob-level proc would save your code from a little bit of copypasting, maybe.

PJB3005 replied:
This code will runtime if you *only* have a single pod connected to the machine when it's destroyed.

Also I fail to see how mob-level proc would help here?

jknpj replied:
good point

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		to_chat(user, "The bluespace conduit flashes violently before calming to a glow.")
+		qdel(A)
+		return
+	if(istype(A, /obj/item/bluespace_crystal/artificial))
```
you have 3 crystals doing the same user.drop_item(A, src)(that needs to be an if() check because the player might not be able to drop the thing), you could simply have a single if for this. checking if it's a flawless/normal/artificial can be done inside that if

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			to_chat(user, "The stone must be empty to function properly.") //prevents deleting players in stones
+			return
+		else
+			user.drop_item(A, src)
```
check if they really dropped the thing

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			soulShardSafety = TRUE
+
+/obj/machinery/mind_machine/mind_machine_hub/attack_hand(user as mob)
+	if(podsConnected == FALSE || connectOne == FALSE || connectTwo == FALSE)
```
```suggestion
	if(!podsConnected || !connectOne || !connectTwo)
```

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+
+/obj/machinery/mind_machine/mind_machine_hub/attack_hand(user as mob)
+	if(podsConnected == FALSE || connectOne == FALSE || connectTwo == FALSE)
+		podsConnected = FALSE
```
might want to give the player some feedback before nuking their stray connections

PJB3005 replied:
Honestly I would go with a button in the UI that is only shown if the connections aren't fulfilled.

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			scanOccupants()
+			return TRUE
+		else
+			errorMessage = "Currently active: Please wait"
```
>keeping a var just for this

Seems like a recipe to disaster

ShiftyRail replied:
Yeah it makes more sense to make it a local var

PJB3005 replied:
I don't see what's wrong with this. Feedback in the UI is much nicer than having it be dropped in the chatbox as long as the message is coded to disappear after like 5 seconds.

GreatBigFailure replied:
This will be the hardest thing to fix and I need more info as to what's wrong with it to make sure I do it right. Do you just want me to have it use to_chat instead? I like having it on the UI but I don't think I'm good enough with nano to avoid a bunch of strings.
As for disappearing after 5 seconds, I intentionally left it to linger because if someone uses a delayed swap and gets sent somewhere I want at least a little evidence on screen of what happened.

JTGSZ commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		connectTwo = null
+	. = ..()
+
+/obj/machinery/mind_machine/mind_machine_hub/attackby(var/obj/item/A as obj, var/mob/user as mob)
```
Remove as obj, and as mob from params. Those are more or less like for verbs to figure shit out, which attackby is not a verb.

JTGSZ replied:
```suggestion
/obj/machinery/mind_machine/mind_machine_hub/attackby(var/obj/item/A, var/mob/user)
	..()
	if(istype(A, /obj/item/bluespace_crystal))
		if(user.drop_item(A, src))
			if(istype(A, /obj/item/bluespace_crystal/artificial))
				bluespaceConduit += 1
				to_chat(user, "The crystal is assimilated into the conduit!")
			else if(istype(A, /obj/item/bluespace_crystal/flawless))
				bluespaceConduit += 300
				to_chat(user, "The bluespace conduit flashes violently before calming to a glow.")
			else
				bluespaceConduit += 3
				to_chat(user, "The crystal is assimilated into the conduit!")
			qdel(A)
			return
		else
			return

	if(istype(A, /obj/item/device/soulstone))
		if(soulShardSafety != FALSE)
			to_chat(user, "That slot is full!")
			return
		if(A.contents.len)
			to_chat(user, "The stone must be empty to function properly.") //prevents deleting players in stones
			return
		else if(user.drop_item(A, src))
			to_chat(user, "The stone is pulled into the machine. You can hear crunching and a quiet sizzle from inside.")
			soulShardSafety = TRUE
			qdel(A)
```
Idk if this is the rite way, I read jknpj's below comment. The as obj/ as mob need cleaned out, and you can save on some checks based on orderin em too i guess.

JTGSZ commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			qdel(A)
+			soulShardSafety = TRUE
+
+/obj/machinery/mind_machine/mind_machine_hub/attack_hand(user as mob)
```
Remove as mob
```suggestion
/obj/machinery/mind_machine/mind_machine_hub/attack_hand(mob/user)
```

JTGSZ commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		occupantTwo = null
+		unlockPods()
+		return
+	if(bluespaceConduit == 0)
```
```suggestion
	if(bluespaceConduit <= 0)
```
you can also do !bluespaceConduit if you are sure it won't go below 0, but the above will handle the scenario where someone somehow goes into the negatives on the number, wherein it will pass that check. Ithinks

PJB3005 replied:
Do not do the `!bluespaceConduit` thing, ever. `bluespaceConduit` is a number, not a boolean. Using boolean operators like not can make it confusing what it actually is because DM isn't strongly typed.

JTGSZ commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		podNumber = 0
+		connectedHub = null
+
+/obj/machinery/mind_machine/mind_machine_pod/MouseDropTo(atom/movable/O as mob|obj, mob/user as mob)
```
```suggestion
/obj/machinery/mind_machine/mind_machine_pod/MouseDropTo(atom/movable/O, mob/user)
```

JTGSZ commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		if(2)
+			connectedHub.occupantTwo = null
+
+/obj/machinery/mind_machine/mind_machine_pod/relaymove(mob/user as mob)
```
```suggestion
/obj/machinery/mind_machine/mind_machine_pod/relaymove(mob/user)
```

JTGSZ commented on code/modules/research/designs/boards/machine_medical.dm:
```diff
@@ -162,3 +162,23 @@
 	build_path = /obj/item/weapon/circuitboard/centrifuge
 
+/datum/design/mind_machine_hub
+	name = "Circuit Design (mind_machine_hub)"
```
```suggestion
	name = "Circuit Design (Mind Machine Hub)"
```

JTGSZ commented on code/modules/research/designs/boards/machine_medical.dm:
```diff
@@ -162,3 +162,23 @@
+	build_path = /obj/item/weapon/circuitboard/mind_machine_hub
+
+/datum/design/mind_machine_pod
+	name = "Circuit Design (mind_machine_pod)"
```
```suggestion
	name = "Circuit Design (Mind Machine Pod)"
```

JTGSZ commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	icon_state = "mind_pod_open"
+	desc = "A large pod used for mind transfers. Contains two locking systems: One for ensuring occupants do not disturb the transfer process, and another that prevents lower minded creatures from leaving on their own."
+	density = 1
+	anchored = 1.0
```
```suggestion
	anchored = 1
```
Can also be TRUE if you feelin it

JTGSZ commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		HT.add_spell(S)
+	for(var/spell/S in HT_spells)
+		HO.add_spell(S)
+	HO.confused = 8
```
```suggestion
	HO.confused += 8
	HO.dizziness += 8
	HT.confused += 8
	HT.dizziness += 8
```
Could add to their confusion and dizziness instead of settin it flat to 8, in the scenario where like they have a shitload and enter the machine to reduce it entirely, I guess?

JTGSZ commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		swapProgress = 0
+		unlockPods()
+
+/obj/machinery/mind_machine/mind_machine_hub/emp_act()
```
```suggestion
/obj/machinery/mind_machine/mind_machine_hub/emp_act(severity)
	..()
	if(currentlySwapping)
		malfSwap = TRUE
		return
```
Machine parent does shit too unless you specifically didn't want it to do the pulse overlay effect etc

JTGSZ commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		HO.remove_spell(S)
+	for(var/spell/S in HT.spell_list)
+		HT.remove_spell(S)
+	var/mob/living/dummy = new(src.loc)
```
```suggestion
	var/mob/living/dummy = new(src)
```
You can avoid having the dummy ever actually appear in any form in realspace and save a bit since you are deleting it by spawning it into the machine itself. I feel liek there may be a better way to save on the spell transfers but idk things well re: that

JTGSZ commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		soulShardSafety = FALSE
+		errorMessage = "Warning: Unknown internal reaction"
+	currentlySwapping = TRUE
+	bluespaceConduit -= 1
```
I saw a += on monkeymen, so I looked for this. are you completely sure you want to like uhhh, use up their bluespace resources that early?

PJB3005 replied:
Indeed. Unless the intent is that the bluespace resources *can* get wasted if it gets interrupted (because there seem to be plenty causes of interruption possible) then you should probably deduct later. There's plenty of situations in which you aren't refunded right now and I'm not sure that's intentional.

Also the lower lifeform thing should probably be an earlier check, yeah?

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	if(istype(A, /obj/item/bluespace_crystal/artificial))
+		user.drop_item(A, src)
+		bluespaceConduit += 1
+		to_chat(user, "The crystal is assimilated into the conduit!")
```
Wrap these in `<span class='notice'></span>`, please.

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			to_chat(user, "That slot is full!")
+			return
+		if(A.contents.len)
+			to_chat(user, "The stone must be empty to function properly.") //prevents deleting players in stones
```
Ditto

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			return
+		else
+			user.drop_item(A, src)
+			to_chat(user, "The stone is pulled into the machine. You can hear crunching and a quiet sizzle from inside.")
```
Ditto

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+/obj/machinery/mind_machine
+	name = "Mind Machine"
```
```suggestion
	name = "\improper Mind Machine"
```

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	machine_flags = SCREWTOGGLE | CROWDESTROY | EMAGGABLE | FIXED2WORK | WRENCHMOVE
+
+/obj/machinery/mind_machine/mind_machine_hub
+	name = "Mind Machine Hub"
```
```suggestion
	name = "\improper Mind Machine Hub"
```

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+				return TRUE
+	if(href_list["eject_pods"])
+		if((!currentlySwapping) && (lockedPods == FALSE))
+			spawn(5)
```
Check if the thing is `gcDestroyed` after the `spawn()`

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			errorMessage = "Currently active: Please wait"
+			return TRUE
+		else
+			spawn(5)
```
Check if the thing is `gcDestroyed` after the `spawn()`

PJB3005 replied:
Returning from spawn like this doesn't work correctly. Either way you should definitely have player feedback that the button *did* something (instead of basically causing artificial UI lag).

If you still want to have this take a bit of time you can call `nanomanager.update_uis(src)` to force update the UI externally.

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			return TRUE
+	if(href_list["mind_swap"])
+		if((occupantScan) && (!currentlySwapping))
+			spawn(5)
```
Check if the thing is `gcDestroyed` after the `spawn()`

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	connectTwo.icon_state = "mind_pod_open"
+	playsound(connectTwo, 'sound/machines/door_unbolt.ogg', 35, 1)
+	playsound(connectOne, 'sound/machines/door_unbolt.ogg', 35, 1)
+	if(badSwap == FALSE)
```
```suggestion
	if(!badSwap)
```

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+/obj/machinery/mind_machine/mind_machine_hub/proc/delayedSwap()
+	if(!occupantScan)
+		errorMessage = "Proceed with caution: supervisor advised"
+		sleep(5 SECONDS)
```
Same thing, please check if it's `gcDestroyed` after the sleep.

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		errorMessage = "Proceed with caution: supervisor advised"
+		sleep(5 SECONDS)
+		currentlySwapping = FALSE
+		if((occupantOne) && (occupantTwo))
```
Pure style, but those extra parenthesis aren't really useful here

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	S = occupantOne
+	occupantNameOne = S.name
+	switch(S.stat)
+		if(0)
```
```suggestion
		if(DEAD)
```

And other defines

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			mindTypeOne = "Artificial" //Silicon player, obviously
+		else
+			mindTypeOne = "Higher" //Player controlled
+	if(isvampire(S) || isanycultist(S) || ischangeling(S) || ismalf(S))
```
!!!!! metatest inbound !!!!!

GreatBigFailure replied:
Even the mind transfer wiz spell disallows those, they're completely borked, spent days trying to fix them but couldn't. I tried to make it a sort of thematic in-universe thing rather than meta "it don't work". 
In fairness every other check for them is easier than shoving them in a mind machine.

PJB3005 replied:
Maybe just make them pretend go through with the transfer as normal but then fail with the brain damage thing instead of actually swapping.

People aren't gonna bother using this as a checking thing if it requires waiting forever *twice* *and is error prone*.

GreatBigFailure replied:
I can change it if you want but consider that this requires moderately expensive construction, very demanding research, and 3 machines to be anchored adjacent to each other. It doesn't require a bluespace crystal to scan but if they're not intentionally tossing them in there it would be required to accidentally discover one. The alternative is drag to church, use any chaplain item, or implant. 2 of which are portable. I just don't think it's a viable test at all and it might even lead to a fun exchange where a cultist has to talk their way out of entering the mind machine without raising suspicion.

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		var/mob/living/carbon/T = S
+		if(T.is_wearing_item(/obj/item/clothing/head/tinfoil))
+			mindTypeTwo = "Shielded"
+	S = occupantTwo
+	occupantNameTwo = S.name //and for pod two
+	switch(S.stat)
```
This is a carbon copy of the thing for occupant one. Please make this a proc so that you don't need to copypaste as much.

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			malfSwap = TRUE
+		sleep(1 SECONDS)
+		swapProgress += 1
+	if((!locate(occupantOne,connectOne.contents)) || (!locate(occupantTwo, connectTwo.contents))) //10 whole seconds for jaunt to cooldown
```
Also check if the machine itself hasn't been `gcDestroyed`

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+
+/obj/machinery/mind_machine/mind_machine_hub/attackby(var/obj/item/A as obj, var/mob/user as mob)
+	..()
+	if(istype(A, /obj/item/bluespace_crystal/flawless))
```
At least do like a `playsound(src, 'sound/weapons/Genhit.ogg', 50, 1)` here for audible feedback.

ShiftyRail commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		return
+	if(!ismob(O))
+		return
+	if(O.loc == user || !isturf(O.loc) || !isturf(user.loc) || !user.Adjacent(O)) //I stole all this from DNA scanners
```
```suggestion
	if(!isturf(O.loc) || !isturf(user.loc) || !user.Adjacent(O)) //I stole all this from DNA scanners
```

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	ui.set_auto_update(1)
+
+/obj/machinery/mind_machine/mind_machine_hub/Topic(href, href_list)
+	if(href_list["scan_start"])
```
All of these will nullref if the user manages to hit the button after a pod has been deconstructed or other such shenanigans. Check whether all pods are connected before processing topic further please.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+
+/obj/machinery/mind_machine/mind_machine_hub/proc/delayedSwap()
+	if(!occupantScan)
+		errorMessage = "Proceed with caution: supervisor advised"
```
This message will (probably) not show up because you're not instructing NanoUI to actually update the UI. Use `nanomanager.update_uis(src)` here before sleeping.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			swapOccupants()
+			unlockPods()
+		else
+			errorMessage = "Missing occupant"
```
Ditto about message not showing up properly.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		return
+	currentlySwapping = TRUE
+	lockedPods = TRUE
+	if((connectOne.icon_state != "mind_pod_closed") && (connectTwo.icon_state != "mind_pod_closed"))
```
Checking icon states to check "were the pods already locked" is a really bad idea. You should keep track of a var on them or something.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		else
+			errorMessage = "Missing occupant"
+
+/obj/machinery/mind_machine/mind_machine_hub/proc/scanOccupants(var/mob/S)
```
Why even have this `S` as a parameter? It's literally only used as a local temporary local variable.

PJB3005 replied:
Also, a ton of the code in this method is duplicated between pod 1 and pod 2. Please unify that.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+			mindTypeOne = "Artificial" //Silicon player, obviously
+		else
+			mindTypeOne = "Higher" //Player controlled
+	if(isvampire(S) || isanycultist(S) || ischangeling(S) || ismalf(S))
```
This feels like an *amazing* antag detector.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	if(bluespaceConduit == 0)
+		errorMessage = "Conduit requires charge"
+		return
+	if(mindTypeOne == "Shielded" || mindTypeTwo == "Shielded")
```
Since these variables are used for actual logic you should make the values into `#define` constants and compare those instead of copy pasting the strings around.

Right now you run the risk of somebody changing the scan code to change the wording and then forgetting to update this check.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		spark(connectTwo)
+		unlockPods()
+		return
+	if(occupantStatOne == "Dead" || occupantStatTwo == "Dead")
```
Here's a fun problem: what if the mob dies while *inside* the pod. Then this variable won't have updated correctly because it's pulled only in the scan proc.

GreatBigFailure replied:
That was actually intentional. Since it already scanned the mind I figured it would make sense to be able to transfer it out. It would also work as a sort of ghetto DNA disk as long as the pod wasn't touched in the meantime.

PJB3005 replied:
May want to leave some sort of comment then to say "yes, this intentional".

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	playsound(connectTwo, 'sound/effects/sparks4.ogg', 80, 1)
+	occupantScan = TRUE
+	S = occupantOne
+	occupantNameOne = S.name
```
Wouldn't it make more sense to check for the mind name instead of the mob display name?

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	for(var/mob/living/simple_animal/borer/B in borers) //borers in the body are counted
+		theFly += B
+	M = occupantTwo
+	for(var/obj/item/weapon/holder/i in get_contents_in_object(M))
```
Again, please de-duplicate this code between occupant one and two.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		if(stat & NOPOWER)
+			malfSwap = TRUE
+		sleep(1 SECONDS)
+		swapProgress += 1
```
You're not calling `nanomanager.update_uis(src)` here so the UI won't update correctly.

GreatBigFailure replied:
I have it set to update automatically as long as it's open. In regard to UI updates that all definitely works.

PJB3005 replied:
The problem with that is it still can cause a maximum of 2 seconds of latency before the UI updates, because that's how slow NanoUI's auto update is.

In this case I guess that's fine, but in a bunch of the others it isn't and can cause noticeable, painful UI lag.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	for(var/mob/living/simple_animal/borer/B in borers)
+		theFly += B
+	for(var/prog in 1 to 40/manipRating) //Counts up 5 to 10 seconds, checks if we lost power each time and ruins your day if so. Otherwise just for UI progress bar
+		if(stat & NOPOWER)
```
I feel like abort checks like "the pods went empty" "the pods got destroyed" etc should be checked in here.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	HT.dizziness = 8
+	currentlySwapping = FALSE
+	occupantScan = FALSE
+	connectOne.icon_state = "mind_pod_closed"
```
You should not duplicate all these "reset state" things all over the place like you are currently.

GreatBigFailure replied:
Do you mean the dizziness = 8 instead of += or things like occupantScan = FALSE?
The latter will be difficult to make into a re-usable proc because it's not always resetting the same things. Unlock() attempts to serve that purpose but I was only able to use it in some places.

PJB3005 replied:
Stuff like the
```
currentlySwapping = FALSE
occupantScan = FALSE
connectOne.icon_state = "mind_pod_closed"
```

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	icon_state = "mind_hub"
+	swapProgress = 0
+
+/obj/machinery/mind_machine/mind_machine_hub/proc/malfunction(var/mob/living/M)
```
Once again you're just using a parameter as a local variable here. Don't.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+		badSwap = TRUE
+		var/list/randFaction = list("hostile", "neutral", "carp", "necro", "wizard", "syndicate", "cult")
+		var/mob/woopsTarget = null
+		if(mindTypeOne == "Higher" || mindTypeOne == "Artificial")
```
Again, remove duplication between the two.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	icon_state = "mind_hub"
+	swapProgress = 0
+
+/obj/machinery/mind_machine/mind_machine_hub/proc/theFlySwap(var/mob/living/MO, var/mob/living/MT)
```
Remove parameters as locals.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	for(var/obj/item/weapon/holder/D in theFly)
+		for(var/mob/M in D.contents)
+			theFly += M //Gets the mob stored in the holder, adds it to the list
+		D.Destroy() //Turns them from holder to mob
```
Do not call `Destroy` directly like this.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	for(var/mob/living/q in theFly)
+		if(q.mind)
+			mindFly += q
+			theFly -= q
```
`theFly` isn't used for the rest of this method until it's reset so there's no use to explicitly removing items from it like this.

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+				if(2)
+					connectedHub.scanRatingTwo = scanCount
+
+/obj/machinery/mind_machine/mind_machine_pod/Destroy()
```
This needs to notify the pod properly to avoid runtimes (e.g. if the pod is destroyed while actively scanning).

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+	if(L == user)
+		visible_message("[user] climbs into \the [src].")
+	else
+		visible_message("[user] puts [L.name] into \the [src].")
```
Any reason you went with `L.name` directly instead of just `L` here?

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+//Swaps over///////////////
+//Pod stuff//////////
+
+/obj/machinery/mind_machine/mind_machine_pod
```
You should hook `Exited` on these pods to make sure the contained mob gets updated in case the player leaves the pod *somehow* (I think there are items that can teleport you out like this even when it's locked?)

PJB3005 commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,846 @@
+				connectedHub.connectOne = null
+			if(2)
+				connectedHub.connectTwo = null
+		connectedHub.podsConnected = null
```
```suggestion
		connectedHub.podsConnected = FALSE
```

This is a boolean, don't set it to null.

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,856 @@
+	light_power_on = 1
+	machine_flags = SCREWTOGGLE | CROWDESTROY | EMAGGABLE | FIXED2WORK | WRENCHMOVE
+
+#define CURRENTLY_ACTIVE "Currently active: please wait"
+#define PODS_UNLOCKED "Pods already unlocked"
+#define PODS_LOCKED "Pods are locked"
+#define REQ_SCAN "Require scan"
+#define DELAY_CAUTION "Proceed with caution: supervisor advised"
+#define MISSING_OCC "Missing occupant"
+#define CONDUIT_CHARGE "Conduit requires charge"
+#define SHIELD_ERROR "Critical error, aborting"
+#define LIVING_REQUIRED "Living mind required"
+#define SHARD_REACTION "Warning: Unknown internal reaction"
+#define SCANNER_INSUFF "Scanners insufficient"
+#define UNKNOWN_INT "Unknown interruption"
+#define SUM_TOO_LOW "Sum intelligence too low"
+#define DELIVERY_ERROR "Error in mind delivery"
+#define MIND_DAMAGED "Mind sustained damage"
+#define SAFETY_CONSUMED "Safety consumed"
+#define TRANSPORT_ANOM "Mind transport anomaly"
+#define NOERROR "No errors"
+
+#define LOWER "Lower"
+#define HIGHER "Higher"
+#define SILICON "Artificial"
+#define SHIELDED "Shielded"
```
```suggestion
#define MIND_MACHINE_STATUS_CURRENTLY_ACTIVE "Currently active: please wait"
#define MIND_MACHINE_STATUS_PODS_UNLOCKED "Pods already unlocked"
#define MIND_MACHINE_STATUS_PODS_LOCKED "Pods are locked"
#define MIND_MACHINE_STATUS_REQ_SCAN "Require scan"
#define MIND_MACHINE_STATUS_DELAY_CAUTION "Proceed with caution: supervisor advised"
#define MIND_MACHINE_STATUS_MISSING_OCC "Missing occupant"
#define MIND_MACHINE_STATUS_CONDUIT_CHARGE "Conduit requires charge"
#define MIND_MACHINE_STATUS_SHIELD_ERROR "Critical error, aborting"
#define MIND_MACHINE_STATUS_LIVING_REQUIRED "Living mind required"
#define MIND_MACHINE_STATUS_SHARD_REACTION "Warning: Unknown internal reaction"
#define MIND_MACHINE_STATUS_SCANNER_INSUFF "Scanners insufficient"
#define MIND_MACHINE_STATUS_UNKNOWN_INT "Unknown interruption"
#define MIND_MACHINE_STATUS_SUM_TOO_LOW "Sum intelligence too low"
#define MIND_MACHINE_STATUS_DELIVERY_ERROR "Error in mind delivery"
#define MIND_MACHINE_STATUS_MIND_DAMAGED "Mind sustained damage"
#define MIND_MACHINE_STATUS_SAFETY_CONSUMED "Safety consumed"
#define MIND_MACHINE_STATUS_TRANSPORT_ANOM "Mind transport anomaly"
#define MIND_MACHINE_STATUS_NOERROR "No errors"

#define MIND_MACHINE_MIND_LOWER "Lower"
#define MIND_MACHINE_MIND_HIGHER "Higher"
#define MIND_MACHINE_MIND_SILICON "Artificial"
#define MIND_MACHINE_MIND_SHIELDED "Shielded"
```

Keep your defines unique

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,856 @@
+		if(2)
+			connectedHub.occupantTwo = occupant
+		else
+			return
```
Useless else, useless return

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,856 @@
+			connectedHub.unlockPods()
+			return
+	src.go_out(ejector = user)
+	return
```
useless return

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,856 @@
+	if(currentlySwapping)
+		spark(src)
+		malfSwap = TRUE
+		return
```
why do you keep returning

GreatBigFailure replied:
Because 90% of this was written when I barely knew what return did. I now feel a tinge of shame.

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,856 @@
+	..()
+	if(currentlySwapping)
+		malfSwap = TRUE
+		return
```
just

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,856 @@
+	var/podsConnected = FALSE
+	var/mob/living/occupantOne = null
+	var/mob/living/occupantTwo = null
+	var/occupantStatOne = null
+	var/occupantStatTwo = null
+	var/mindTypeOne = null	//Player mind, simple mob, silicon, etc
+	var/mindTypeTwo = null
```
You really don't need these vars since the occupants are typecast.

GreatBigFailure replied:
Those ones I do. I got rid of name for that reason but stat serves a single minor purpose in allowing people who were scanned when alive to still be swapped after death. I like that as a feature and want to keep it. It also admittedly just makes it much easier for me to have the UI display a their actual state rather than a number.

jknpj commented on code/game/machinery/mind_machine.dm:
```diff
@@ -0,0 +1,856 @@
+//////Parts and connection/////////
+
+/obj/machinery/mind_machine/mind_machine_hub/New()
+	illegalSwap = boss_mobs + blacklisted_mobs - list(/mob/living/simple_animal/hostile/mechahitler, /mob/living/simple_animal/hostile/alien/queen/large, /mob/living/simple_animal/hostile/retaliate/cockatrice)
```
>cockatrice

*[foreseeable consequences intensifies]*

GreatBigFailure replied:
I foresee fun
Fun and statues
